### PR TITLE
Resurrecting axes_grid1 documentation

### PR DESCRIPTION
**Pull request by ImportanceOfBeingErnest**
## PR Summary

This is an attempt to resurrect the axes_grid1 and axisartist documentation.
Those are not present in the current docs, see #9582

The strategy here is to move stuff from  
`doc/mpl_toolkits/` → `doc/api/toolkits`
and create a toc for both modules.

Then, a lot of broken links need to be replaced. 

<strike>Currently this still fails due to `mpl_toolkits.axes_grid1.colorbar.colorbar` not producing a valid docstring.
There are also a couple of missing files which should be produced by sphinx-gallery, but are not for some reason. </strike> (solved)

<strike>I will try to include the make.bat into this PR without `-W` flag, such that the tests can hopefully build the docs and people can look at the produced errors.</strike> (solved)

The general idea here would be:
* create a working documentation for axisartist and axes_grid1 inside of `doc/api/toolkits` first.
* As can be seen, this is unproblematic in general, but causes some trouble with broken docstrings and missing files etc. Those have now been solved. In particular:
  * problematic links have been replaced by the full version (`~matplotlib.axes.Axes` instead of `.Axes`)
  * classes for which no examples are created by sphinx-gallery (https://github.com/sphinx-gallery/sphinx-gallery/issues/365) are excluded inside the template.
  * The offending `colorbar` vs `Colorbar` issue (https://github.com/sphinx-doc/sphinx/issues/4874) has been worked around by using a different template for this special case.
* Now that those problems are solved and a valid documentation can be built, one may in a next step move it outside of `doc/api` - in case this is still desired. (I didn't manage to do this, so I would propose to do this in a new PR if someone has a solution to that).

## PR Checklist

- [x] Code is PEP 8 compliant
- [x] Documentation is sphinx and numpydoc compliant

### CODE REVIEW

**Comment by ImportanceOfBeingErnest** [comment]
This is not yet ready for review in the sense of a working code; and I suppose tests will fail. But the docs can be built without `-W` flag, i.e. `SPHINXOPTS=`. In that sense, if anyone wants to comment on this, they should. 

The problem of keeping the doc/mpl_toolkits is that what is in there can hardly be synchronized with the rest of the docs. While if you put it into doc/api/... it can live there without problem.

File: tutorials/text/text_intro.py [hunk 1]
```diff
@@ -38,17 +38,20 @@
 `.pyplot` API       OO API              description
 =================== =================== ======================================
 `~.pyplot.text`     `~.Axes.text`       Add text at an arbitrary location of
-                                        the `.Axes`.
+                                        the `~matplotlib.axes.Axes`.
```
Comment (dstansby) [inline tutorials/text/text_intro.py]: You should be able to do `~.axes.Axes` that saves a bit of space (and similarly for the changes below)

Comment (ImportanceOfBeingErnest) [reply tutorials/text/text_intro.py]: Probably. But due to three different `Axes` being around I wanted to be on the safe side first. ;-)

**Comment by ImportanceOfBeingErnest** [comment]
The problem about missing files seems to be due to sphinx-gallery not creating `.examples` files for aliased classes. I opened issue https://github.com/sphinx-gallery/sphinx-gallery/issues/365 over there.

**Comment by ImportanceOfBeingErnest** [comment]
The colorbar problem is however a real mystery. If I rename the `mpl_toolkits.axes_grid1.colorbar.colorbar` method to `donaldduck`, the respective doc page are created without problem (this just breaks some of the examples of course).

**Comment by ImportanceOfBeingErnest** [comment]
The last blocker here is that it is apparently not possible to get an autodoc/autosummary documentation created for a file that contains a class and a function of the same name, one uppercase, one lowercase. 

```
class Colorbar():
      # ...

def colorbar():
     # ...
```

I opened this issue https://github.com/sphinx-doc/sphinx/issues/4874 about it.

**Comment by anntzer** [comment]
I bet the issue will be closed as wontfix on sphinx, as it's a fundamental problem of the Windows FS (that it doesn't distinguish (*modulo really low-level hacks) files differing only based case; and any fix I can think of would involve breaking the correspondence [pythonname] -> [pythonname.rst] -> [pythonname.html], which seems to be quite a big price to pay).

Looks like the stuff builds fine on circleci, so that's good enough for me.

File: doc/_templates/autosummary.rst [hunk 2]
```diff
@@ -6,11 +6,14 @@
 .. auto{{ objtype }}:: {{ objname }}
 
 {% if objtype in ['class', 'method', 'function'] %}
+{% if objname not in ['AxesGrid', 'Scalable', 'HostAxes', 'FloatingAxes',
+                      'ParasiteAxesAuxTrans', 'ParasiteAxes'] %}
```
Comment (anntzer) [inline doc/_templates/autosummary.rst]: I assume the plan is to remove these after their docstrings are fixed?  Perhaps leave a comment to that effect.

File: doc/api/toolkits/axes_grid.rst [hunk 3]
```diff
@@ -20,9 +20,8 @@ Axes Grid
...
    <https://matplotlib.org/2.0.1/mpl_toolkits/axes_grid/index.html#toolkit-axesgrid-index>`_.
 
-.. autosummary::
-   :toctree: ../_as_gen
-   :template: automodule.rst
 
-   axes_grid1
-   axisartist
+
+:ref:`toolkit_axisartist-index`
```
Comment (anntzer) [inline doc/api/toolkits/axes_grid.rst]: what happened here?

File: doc/make.bat [hunk 4]
```diff
@@ -10,7 +10,7 @@ if "%SPHINXBUILD%" == "" (
 set SOURCEDIR=.
 set BUILDDIR=build
 set SPHINXPROJ=matplotlib
-set SPHINXOPTS=-W
+set SPHINXOPTS=
```
Comment (anntzer) [inline doc/make.bat]: needs to come back

File: doc/_templates/autosummary.rst (hunk 2 above)
Comment (ImportanceOfBeingErnest) [reply doc/_templates/autosummary.rst]: This relates to sphinx-gallery/sphinx-gallery#365; so yes, once that is fixed, one can remove.

File: doc/make.bat (hunk 4 above)
Comment (ImportanceOfBeingErnest) [reply doc/make.bat]: Sure. I just need this for testing locally.

**Comment by anntzer** [comment]
What do you get on Windows? A complete failure to build (even removing -W from sphinxopts), or just something wrong wrt colorbar links?
If the latter, I wouldn't consider that a blocker because solving casefolding issues on Windows is well beyond what Matplotlib should be worried about (see e.g. the issues for git and mercurial on that same topic).  If the former I'd be more agreeable to waiting for a fix.

**Comment by jklymak** [comment]
After the config changes get reverted, this is great.  Great job @ImportanceOfBeingErnest this is a huge improvement!

**Comment by ImportanceOfBeingErnest** [comment]
Without `-W` flag, i.e. turning errors into warnings, you get a lot of red lines in the console when building but the output is fine, except for the .colorbar module and any links to or from it.   
In that sense maybe you're right and one would just have to include a sentence about removing `-W` on the page about building the docs on windows.

**Comment by ImportanceOfBeingErnest** [comment]
I will try to get a good version of this together tomorrow - there is still a lot of pieces to glue together.

**Comment by anntzer** [comment]
I would just remove the -W from make.bat then, with a comment explaining why.

To be clear, I think Windows support is just as important as Linux/OSX (see https://github.com/matplotlib/matplotlib/issues/3148 for literally my first issue (not PR) on this repo), but the specific issue at hand is so complex that it's just not worth blocking on it.

File: doc/make.bat (hunk 4 above)
Comment (dstansby) [reply doc/make.bat]: You can do `make html "SPHINXOPTS="` to override the `-W` flag locally I think.

**Comment by ImportanceOfBeingErnest** [comment]
I think I found an acceptable workaround, not using the autosummary for the colorbar docs at all. This allows to keep the `-W` flag. 

I think this should now be ready for review.

File: lib/matplotlib/pyplot.py [hunk 5]
```diff
@@ -898,16 +898,16 @@ def subplot(*args, **kwargs):
...
+    at position *index* of a (virtual) grid of *nrows* by *ncols* axes.
+    Indexes go from 1 to ``nrows * ncols``, incrementing in row-major order.
 
     If *nrows*, *ncols* and *index* are all less than 10, they can also be
     given as a single, concatenated, three-digit number.
 
     For example, ``subplot(2, 3, 3)`` and ``subplot(233)`` both create an
-    `.Axes` at the top right corner of the current figure, occupying half of
-    the figure height and a third of the figure width.
+    `matplotlib.axes.Axes` at the top right corner of the current figure,
```
Comment (jklymak) [inline lib/matplotlib/pyplot.py]: In future, it'd be nicer if changes like this that are orthogonal to the main PR are put in a separate PR.  Did you need to do these to get the docs to build or something?

File: tutorials/intermediate/constrainedlayout_guide.py [hunk 6]
```diff
@@ -543,10 +543,11 @@ def docomplicated(suptitle=None):
 # Each item has a layoutbox associated with it. The nesting of gridspecs
 # created with `.GridSpecFromSubplotSpec` can be arbitrarily deep.
 #
-# Each `.Axes` has *two* layoutboxes.  The first one ``ax._layoutbox``
-# represents the outside of the Axes and all its decorations (i.e. ticklabels,
-# axis labels, etc.).  The second layoutbox corresponds to the Axes'
-# `ax.position`, which sets where in the figure the spines are placed.
+# Each ``~matplotlib.axes.Axes` has *two* layoutboxes.  The first one,
```
Comment (jklymak) [inline tutorials/intermediate/constrainedlayout_guide.py]: Again, why did you change these?  The original worked just fine in the main docs...  https://matplotlib.org/tutorials/intermediate/constrainedlayout_guide.html#sphx-glr-tutorials-intermediate-constrainedlayout-guide-py

File: tutorials/text/text_intro.py (hunk 1 above)
Comment (jklymak) [reply tutorials/text/text_intro.py]: Umm, OK, so does this change *break* the simple `.Axes` shortcut?  If so, I'm a little leery about accepting it, because I can't imagine there aren't other places in the code where this shortcut has been used then the ones you've flagged above....

**Comment by jklymak** [comment]
Milestoning 3.0, but happy for it to be back ported if possible...

**Comment by ImportanceOfBeingErnest** [comment]
Yes, `.Axes` wont work any more. But I suppose `~.axes.Axes` would (as commented by @dstansby). I could change them all to `~.axes.Axes` if people want that.  
I did change *all* occurences, so you don't need to be afraid of that. Essentially if we keep the `-W` flag active, that will be an automatic test. If someone puts `~.Axes` somewhere the doc built would fail with a rather self-explanatory error in how far this is ambiguous.

**Comment by ImportanceOfBeingErnest** [comment]
@tacaswell This is the PR I was refering to, which revives the `axes_grid1` and `axisartist` documentation. The respective docs had been present in v2.0.2 see [e.g. this page](https://matplotlib.org/mpl_toolkits/axes_grid/api/index.html#the-matplotlib-axes-grid1-toolkit-api)

image

But then it has disappeared, such that the [current 2.2.2 version](https://matplotlib.org/api/_as_gen/mpl_toolkits.axes_grid1.html#module-mpl_toolkits.axes_grid1) looks pretty empty. 

image

This PR has revived the documentation. In that sense it fixes a regression. One could therefore argue that it should be backported.  
Of course one could also argue that it is non-critical because there is still the 2.0.2 version available from the internet in case people need it.

**Comment by tacaswell** [comment]
@meeseeksdev backport to v2.2.x

**Comment by tacaswell** [comment]
This can't go back to 2.2.2-doc due to the changes to the source files, lets see if it backports cleanly...
